### PR TITLE
TS aggregation using xarray

### PR DESCRIPTION
**Pull request by diodon**
TSaggr2 is the aggregator function using xarray+pandas. It uses:
- json config file for the global attrs (TSagg_globalmetadata.json)
- json config file for tha variable attrs (TSagg_variableAttributes.json)
- json config file for the input parameter of the function (TSaggr_config.json)

It depends on the geoserverCatalog::get_mooring_url to harvest the urls form the geoserver.

It runs 3x faster and the code is 2x shorter and cleaner. The string variables area `dtype=str` not a char array. This is not CF compliant, but it is easier to handle. It is also very easy to change back to a char array if needed.

### CODE REVIEW

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	{"axis": "Z",
+	 "long_name": "nominal depth",
+	 "positive": "down",
+	 "reference_datum": "sea surface",
+	 "standard_name": "depth",
+	 "units": "m",
+	 "valid_max": 12000.0,
+	 "valid_min": -5.0},
+	 
+"INSTRUMENT_ID":
```
Comment (mhidas) [inline ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: Rename to "instrument_index"

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	 "standard_name": "depth",
+	 "units": "m",
+	 "valid_max": 12000.0,
+	 "valid_min": -5.0},
+	 
+"INSTRUMENT_ID": 
+	{"long_name": "which instrument this obs is for", 
+	 "instance_dimension": "INSTRUMENT"},
+	 
+"INSTRUMENT_TYPE":
```
Comment (mhidas) [inline ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: "instrument_id"

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	{"long_name": "source deployment code, instrument make, model, serial_number"},
+	
+"FILENAME":
```
Comment (mhidas) [inline ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: "source_file"

**Comment by mhidas** [comment]
Also, please rename TSaggr2.py to aggregated_timeseries.py, and remove copyDataset.py to avoid confusion.

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+
+## create empty DF with multiple dtypes
+def create_empty_dataframe(columns):
+    #index = pd.Index([], name=columns[0][0], dtype=columns[0][1])
+    # create the dataframe from a dict
+    return pd.DataFrame({k: pd.Series(dtype=t) for k, t in columns})
+
+
+def main_aggregator(files_to_agg, var_to_agg):
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                 + "_" + split_parts[4] \
+                 + "_" + "FV0" + str(file_version) \
+                 + "_" + VoI + "_" + file_product_type + "-aggregate" \
+                 + "_END-" + t_end \
+                 + "_C-" + datetime.utcnow().strftime(file_timeformat) \
+                 + ".nc"
+    return output_name
+
+def create_empty_dataframe(columns):
+    #index = pd.Index([], name=columns[0][0], dtype=columns[0][1])
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    variableAuxDF = create_empty_dataframe(AuxDF_types)
+
+    ## main loop
+    fileIndex = 0
+    for file in files_to_agg:
+        print(fileIndex, end=" ")
+        sys.stdout.flush()
+
+
+        #with xr.open_dataset(file, use_cftime=True) as nc:
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json (hunk 1 above)
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: done

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json (hunk 2 above)
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: done

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json (hunk 3 above)
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: done

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    # create the dataframe from a dict
+    return pd.DataFrame({k: pd.Series(dtype=t) for k, t in columns})
+
+def main_aggregator(files_to_agg, var_to_agg):
+
+    ## constants
+    UNITS = 'days since 1950-01-01 00:00 UTC'
+    CALENDAR = 'gregorian'
+    FILLVALUE = 999999.0
+    FILLVALUEqc = 99
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: QC set flag set to 9 in case of DEPTH is not present

[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: _FillValue False for QC variables

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    #index = pd.Index([], name=columns[0][0], dtype=columns[0][1])
+    # create the dataframe from a dict
+    return pd.DataFrame({k: pd.Series(dtype=t) for k, t in columns})
+
+def main_aggregator(files_to_agg, var_to_agg):
+
+    ## constants
+    UNITS = 'days since 1950-01-01 00:00 UTC'
+    CALENDAR = 'gregorian'
+    FILLVALUE = 999999.0
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+            DF = DF[(DF['TIME']>=deploymentStart) & (DF['TIME']<=deploymentEnd)]
+
+            ## append data
+            variableMainDF = pd.concat([variableMainDF, DF], ignore_index=True)
+
+            # append auxiliary data
+            # this could be more efficient if I store the variables in a list and make one concat at the end
+            variableAuxDF = variableAuxDF.append({'FILENAME': file,
+                                                  'PLATFORM_CODE': nc.attrs['platform_code'],
+                                                  'INSTRUMENT_TYPE': nc.attrs['deployment_code'] + '-' + nc.attrs['instrument'] + '-' + nc.attrs['instrument_serial_number'],
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: "; " as separator

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+            ## select only in water data
+            DF = DF[(DF['TIME']>=deploymentStart) & (DF['TIME']<=deploymentEnd)]
+
+            ## append data
+            variableMainDF = pd.concat([variableMainDF, DF], ignore_index=True)
+
+            # append auxiliary data
+            # this could be more efficient if I store the variables in a list and make one concat at the end
+            variableAuxDF = variableAuxDF.append({'FILENAME': file,
+                                                  'PLATFORM_CODE': nc.attrs['platform_code'],
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: removed

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                'geospatial_lat_min':       nc.LATITUDE.values.min(),
+                'geospatial_lat_max':       nc.LATITUDE.values.max(),
+                'geospatial_lon_min':       nc.LONGITUDE.values.min(),
+                'geospatial_lon_max':       nc.LONGITUDE.values.max(),
+                'date_created':             datetime.utcnow().strftime(timeformat),
+                'history':                  datetime.utcnow().strftime(timeformat) + ': Aggregated file created.',
+                'keywords':                 ', '.join(list(nc.variables) + ['AGGREGATED']),
+                'lineage':                  'The variable of interest (VoI) is produced by sequentially concatenating the individual values in each of the input files. ' + \
+                                            'The resulting variable has dimension OBS. The VoI’s ancillary_variables, in particular the corresponding quality-control flags, are also included, with dimension OBS. ' + \
+                                            'If the quality control variable is absent in any input file, the corresponding flags in the output file will be set to 0 (“no QC performed”). ' + \
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: fixed

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	 "long_name": "longitude",
+	 "reference_datum": "WGS84 geographic coordinate system",
+	 "standard_name": "longitude",
+	 "units": "degrees_east",
+	 "valid_max": 180.0,
+	 "valid_min": -180.0},
+	 
+"DEPTH": 
+	{"ancillary_variables": "DEPTH_quality_control",
+	 "comment": "depthPP: Depth inferred from 2 neighbouring pressure sensors WET Labs WQM @21m and WET Labs WQM @45m, using the Gibbs-SeaWater toolbox (TEOS-10) v3.06 from latitude and relative pressure measurements (calibration offset usually performed to balance current atmospheric pressure and acute sensor precision at a deployed depth).",
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: deleted

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	 "valid_min": -5.0},
+	 
+"DEPTH_quality_control": 
+	{"_Unsigned": "false",
+	 "flag_meanings": "No_QC_performed Good_data Probably_good_data Bad_data_that_are_potentially_correctable Bad_data Value_changed Not_used Not_used Not_used Missing_value",
+	 "flag_values": "0, 1, 2, 3, 4, 5, 6, 7, 8, 9",
+	 "long_name": "quality flag for depth",
+	 "quality_control_conventions": "IMOS standard flags",
+	 "quality_control_global": "B",
+	 "quality_control_global_conventions": "Argo reference table 2a (see http://www.cmar.csiro.au/argo/dmqc/user_doc/QC_flags.html), applied on data in position only (between global attributes time_deployment_start and time_deployment_end)",
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: both deleted

File: ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+	 "long_name": "actual depth",
+	 "positive": "down",
+	 "reference_datum": "sea surface",
+	 "standard_name": "depth", 
+	 "units": "m",
+	 "valid_max": 12000.0,
+	 "valid_min": -5.0},
+	 
+"DEPTH_quality_control": 
+	{"_Unsigned": "false",
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_variableAttributes.json]: deleted

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    #files_to_aggregate = ['http://thredds.aodn.org.au/thredds/dodsC/IMOS/ANMN/QLD/GBRMYR/Biogeochem_timeseries/IMOS_ANMN-QLD_CKOSTUZ_20161121T005927Z_GBRMYR_FV01_GBRMYR-1611-WQM-195_END-20170606T080027Z_C-20170626T052818Z.nc']
+
+    nc = main_aggregator(files_to_agg=files_to_aggregate, var_to_agg=varname)
+
+    ## set global attributes
+    globalattr_file = 'TSagg_globalmetadata.json'
+    nc.attrs = set_globalattr(nc, globalattr_file, varname, site)
+
+
+    ncout_filename = generate_netcdf_output_filename(fileURL=files_to_aggregate[0], nc=nc, VoI=varname, file_product_type='Full', file_version=1)
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: chenged

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+                          'DEPTH_quality_control':          (['ObservationIndex'],variableMainDF['DEPTH_quality_control'].astype(np.byte), variable_attributes['DEPTH_quality_control']),
+                          'instrument_index':               (['ObservationIndex'],variableMainDF['INSTRUMENT_ID'].astype('int64'), variable_attributes['INSTRUMENT_ID']),
+                          'LONGITUDE':                      (['InstrumentIndex'], variableAuxDF['LONGITUDE'].astype('float32'), variable_attributes['LONGITUDE']),
+                          'LATITUDE':                       (['InstrumentIndex'], variableAuxDF['LATITUDE'].astype('float32'), variable_attributes['LATITUDE']),
+                          'NOMINAL_DEPTH':                  (['InstrumentIndex'], variableAuxDF['NOMINAL_DEPTH']. astype('float32'), variable_attributes['NOMINAL_DEPTH']),
+                          'instrument_id':                  (['InstrumentIndex'], variableAuxDF['INSTRUMENT_TYPE'].astype('str'), variable_attributes['INSTRUMENT_TYPE'] ),
+                          'source_file':                    (['InstrumentIndex'], variableAuxDF['FILENAME'].astype('str'), variable_attributes['FILENAME'])})
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: Done, although I think `Observationindex` and `Instrumentindex` are more informative. I suggest to use those and change it in the IP

```
netcdf IMOS_ANMN-NRS_TZ_20141216T080000Z_NRSROT_FV01_TEMP_aggregated-time-series_END-20150330T013000Z_C-20190705 {
dimensions:
        OBS = 119354 ;
        INSTRUMENT = 4 ;
```

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+            variableMainDF = pd.concat([variableMainDF, DF], ignore_index=True)
+
+            # append auxiliary data
+            # this could be more efficient if I store the variables in a list and make one concat at the end
+            variableAuxDF = variableAuxDF.append({'FILENAME': file,
+                                                  'PLATFORM_CODE': nc.attrs['platform_code'],
+                                                  'INSTRUMENT_TYPE': nc.attrs['deployment_code'] + '-' + nc.attrs['instrument'] + '-' + nc.attrs['instrument_serial_number'],
+                                                  'LONGITUDE': nc.LONGITUDE.squeeze().values,
+                                                  'LATITUDE': nc.LATITUDE.squeeze().values,
+                                                  'NOMINAL_DEPTH': nc.attrs['instrument_nominal_depth']}, ignore_index = True)
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: fixed.

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+                          'instrument_id':                  (['InstrumentIndex'], variableAuxDF['INSTRUMENT_TYPE'].astype('str'), variable_attributes['INSTRUMENT_TYPE'] ),
+                          'source_file':                    (['InstrumentIndex'], variableAuxDF['FILENAME'].astype('str'), variable_attributes['FILENAME'])})
+
+    nc_aggr.TIME.encoding['units'] = UNITS
+    nc_aggr.TIME.encoding['calendar'] = CALENDAR
+    nc_aggr.DEPTH.encoding['_FillValue'] = FILLVALUE
+    nc_aggr.DEPTH_quality_control['_FillValue'] = FILLVALUEqc
+
+
+    return nc_aggr
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: sort added: 
```
    variableMainDF.sort_values(by=['TIME'], inplace=True)
```

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+        print(fileIndex, end=" ")
+        sys.stdout.flush()
+
+
+        #with xr.open_dataset(file, use_cftime=True) as nc:
+        with xr.open_dataset(file) as nc:
+            varnames = list(nc.variables.keys())
+
+            deploymentStart = parse(nc.attrs['time_deployment_start'])
+            deploymentEnd = parse(nc.attrs['time_deployment_end'])
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done
```
            time_deployment_start = pd.to_datetime(parse(nc.attrs['time_deployment_start'])).tz_localize(None)
            time_deployment_end = pd.to_datetime(parse(nc.attrs['time_deployment_end'])).tz_localize(None)
```

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+
+            if 'DEPTH' in varnames:
+                DF = pd.DataFrame({ 'TIME': nc.TIME.squeeze(),
+                                    'VAR': nc[var_to_agg].squeeze(),
+                                    'VARqc': nc[var_to_agg + '_quality_control'].squeeze(),
+                                    'DEPTH': nc.DEPTH.squeeze(),
+                                    'DEPTH_quality_control': nc.DEPTH_quality_control.squeeze(),
+                                    'INSTRUMENT_ID': np.repeat(fileIndex, len(nc['TIME']))})
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: done

File: ANMN/LTSP/TSaggregator/TSagg_globalmetadata.json [hunk 21]
```diff
@@ -0,0 +1,23 @@
+{
+"abstract": "Aggregated Time-series Product: This file contains all measurements of the selected variable from all instruments deployed at the selected site. Timestamps are chronologically ordered, but may not be at uniform intervals. Instrument details are stored as a variable in order to keep a record of the origin of each measurement. The quality control flags of the variable of interest and DEPTH are preserved. Out-of-water measurements have been excluded, but no other filtering has been applied to the input data.", 
+"acknowledgement": "Any users of IMOS data are required to clearly acknowledge the source of the material derived from IMOS in the format: \"Data was sourced from the Integrated Marine Observing System (IMOS) - IMOS is a national collaborative research infrastructure, supported by the Australian Government.\" If relevant, also credit other organisations involved in collection of this particular datastream (as listed in 'credit' in the metadata record).",
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSagg_globalmetadata.json]: single attribute template JSON file. code modified accordingly.

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+def create_empty_dataframe(columns):
+    #index = pd.Index([], name=columns[0][0], dtype=columns[0][1])
+    # create the dataframe from a dict
+    return pd.DataFrame({k: pd.Series(dtype=t) for k, t in columns})
+
+def main_aggregator(files_to_agg, var_to_agg):
+
+    ## constants
+    UNITS = 'days since 1950-01-01 00:00 UTC'
+    CALENDAR = 'gregorian'
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: attribute define in the JSON file

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+            fileIndex += 1
+    print()
+
+    ## rename indices
+    variableAuxDF.index.rename('InstrumentIndex', inplace=True)
+    variableMainDF.index.rename('ObservationIndex', inplace=True)
+
+    ## get variable attributes
+    variable_attributes_templatefile = 'TSagg_variableAttributes.json'
+    variable_attributes = set_variableattr(nc, var_to_agg, variable_attributes_templatefile)
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: All possible variables attributes (only the variables specified in the business case) are now defined in the JSON file.

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+    ## set encoding for netCDF file
+    encoding = {'TIME':                     {'_FillValue': False,
+                                             'units': 'days since 1950-01-01 00:00 UTC',
+                                             'calendar': 'gregorian'},
+                'LATITUDE':                 {'_FillValue': False},
+                'LONGITUDE':                {'_FillValue': False},
+                'DEPTH':                    {'_FillValue': 999999.0},
+                'DEPTH_quality_control':    {'_FillValue': 99},
+                varname:                    {'_FillValue': 999999.0},
+                varname+'_quality_control': {'_FillValue': 99}}
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: in JSON file now.

File: ANMN/LTSP/TSaggregator/TSaggr2.py [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+    t_end = pd.to_datetime(nc.TIME.min().values).strftime(nc_timeformat)
+
+    split_path = fileURL.split("/")
+    split_parts = split_path[-1].split("_") # get the last path item (the file nanme)
+
+    output_name = split_parts[0] + "_" + split_parts[1] + "_" + split_parts[2] \
+                 + "_" + t_start \
+                 + "_" + split_parts[4] \
+                 + "_" + "FV0" + str(file_version) \
+                 + "_" + VoI + "_" + file_product_type + "-aggregate" \
```
[the remark this answers is not in the document]
Comment (diodon) [reply ANMN/LTSP/TSaggregator/TSaggr2.py]: tidier generator function

**Pull request merged**